src/plugins/discover/public/plugin.ts

```diff
@@ -99,24 +152,171 @@ export class DiscoverPlugin implements Plugin<DiscoverSetup, DiscoverStart> {
+        await this.initializeInnerAngular();
+
+        // make sure the index pattern list is up to date
+        await dataStart.indexPatterns.clearCache();
+        const { renderApp } = await import('./application/application');
+        const unmount = await renderApp(innerAngularName, params.element);
+        return () => {
+          unmount();
+          appUnMounted();
+        };
+      },
+    });
+
+    if (plugins.home) {
+      registerFeature(plugins.home);
+    }
+
+    this.registerEmbeddable(core, plugins);
+
     return {
       docViews: {
         addDocView: this.docViewsRegistry.addDocView.bind(this.docViewsRegistry),
-        setAngularInjectorGetter: this.docViewsRegistry.setAngularInjectorGetter.bind(
-          this.docViewsRegistry
-        ),
       },
     };
   }
 
-  start() {
+  start(core: CoreStart, plugins: DiscoverStartPlugins) {
+    // we need to register the application service at setup, but to render it
+    // there are some start dependencies necessary, for this reason
+    // initializeInnerAngular + initializeServices are assigned at start and used
+    // when the application/embeddable is mounted
+    this.initializeInnerAngular = async () => {
+      if (this.innerAngularInitialized) {
+        return;
+      }
+      // this is used by application mount and tests
+      const module = getInnerAngularModule(
+        innerAngularName,
+        core,
+        plugins,
+        this.initializerContext
+      );
+      setAngularModule(module);
+      this.innerAngularInitialized = true;
+    };
+
+    this.initializeServices = async () => {
+      if (this.servicesInitialized) {
+        return { core, plugins };
+      }
+      const services = await buildServices(core, plugins, this.initializerContext);
+      setServices(services);
+      this.servicesInitialized = true;
+
+      return { core, plugins };
+    };
+
     return {
-      docViews: {
-        DocViewer,
-      },
       savedSearches: {
         createLoader: createSavedSearchesLoader,
       },
     };
   }
+
+  stop() {
+    if (this.stopUrlTracking) {
+      this.stopUrlTracking();
+    }
+  }
+
+  /**
+   * register embeddable with a slimmer embeddable version of inner angular
+   */
+  private async registerEmbeddable(
+    core: CoreSetup<DiscoverStartPlugins>,
+    plugins: DiscoverSetupPlugins
+  ) {
+    const { SearchEmbeddableFactory } = await import('./application/embeddable');
+
+    if (!this.getEmbeddableInjector) {
+      throw Error('Discover plugin method getEmbeddableInjector is undefined');
+    }
+
+    const getStartServices = async () => {
+      const [coreStart, deps] = await core.getStartServices();
+      return {
+        executeTriggerActions: deps.uiActions.executeTriggerActions,
+        isEditable: () => coreStart.application.capabilities.discover.save as boolean,
+      };
+    };
+
+    const factory = new SearchEmbeddableFactory(getStartServices, this.getEmbeddableInjector);
```

Comment:
I agree we should clean those up, but IMHO this can happen in a separate PR. I created an issue to track the way we are passing services around a while ago: https://github.com/elastic/kibana/issues/63596

It also references the `createStartServiceGetter` function.